peer/abstractlist/list.go

```diff
@@ -0,0 +1,592 @@
+func (pl *List) Transport() peer.Transport { return pl.transport }
+
+// Update applies the additions and removals of peer Identifiers to the list
+// it returns a multi-error result of every failure that happened without
+// circuit breaking due to failures.
+func (pl *List) Update(updates peer.ListUpdates) error {
+	if len(updates.Additions) == 0 && len(updates.Removals) == 0 {
+		return nil
+	}
+
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	if !pl.once.IsRunning() {
+		return pl.updateOffline(updates)
+	}
+	return pl.updateOnline(updates)
+}
+
+func (pl *List) updateOnline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.remove(id))
+	}
+
+	add := updates.Additions
+	if !pl.noShuffle {
+		add = shuffle(pl.randSrc, add)
+	}
+
+	for _, id := range add {
+		errs = multierr.Append(errs, pl.add(id))
+	}
+	return errs
+}
+
+func (pl *List) updateOffline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.removeOffline(id))
+	}
+	for _, id := range updates.Additions {
+		errs = multierr.Append(errs, pl.addOffline(id))
+	}
+	return errs
+}
+
+// Start notifies the List that requests will start coming
+func (pl *List) Start() error {
+	return pl.once.Start(pl.start)
+}
+
+func (pl *List) start() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.offlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// Stop notifies the List that requests will stop coming
+func (pl *List) Stop() error {
+	return pl.once.Stop(pl.stop)
+}
+
+func (pl *List) stop() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.onlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// add retains a peer and sets up a thunk (a thin proxy for a peer) to receive
+// connection status notifications from the dialer and track pending request
+// counts.
+//
+// add does not add the peer to the list of peers available for choosing (the
+// Implementation).
+// The thunk is responsible for adding and removing the peer from the
+// collection of available peers based on connection status notifications.
+// Must be run inside a mutex.Lock()
+func (pl *List) add(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	if _, ok := pl.peers[addr]; ok {
+		return peer.ErrPeerAddAlreadyInList(addr)
+	}
+
+	t := newPeerFacade(pl, id)
+	t.boundOnFinish = t.onFinish
+
+	p, err := pl.transport.RetainPeer(id, t)
```

Comment:
Yep. This may be the crux of our deadlock issues.

In #1660, I eliminated this problem for HTTP by ensuring that all notifications are synchronized and emitted from the peer status monitor goroutine. I’ll be doing the same for the gRPC and TChannel transports next.

At the moment, these notifications come from the abstract peer implementation and are triggered by start request, finish request, and dial. All the dialing for all three transports occurs in a goroutine and never on the retain/release stack.

Reply:
Even today, connection status notifications are emitted from a goroutine in every transport. There is actually very little chance that this is the source of our deadlocks. It’s more likely that the deadlock in gRPC was related to the obfuscation I’m addressing in #1834 or possibly a loop between transport.StartRequest/EndRequest and peer list Choose/onFinish, which I’m addressing with this change, #1829, #1830, #1831, #1832, #1835, and #1836.